python_modules/dagster/dagster/_core/definitions/decorators/asset_decorator.py

```diff
@@ -617,6 +618,192 @@ def build_asset_ins(
+            not provided, the name "default" is used.
+
+    Examples:
+        .. code-block:: python
+
+            @op
+            def fetch_files_from_slack(context) -> pd.DataFrame:
+                ...
+
+            @op
+            def store_files_in_table(files) -> None:
+                files.to_sql(name="slack_files", con=create_db_connection())
+
+            @graph_asset
+            def slack_files_table():
+                return store_files(fetch_files_from_slack())
+    """
+    if compose_fn is not None:
+        return _GraphBackedAsset()(compose_fn)
+
+    def inner(fn: Callable[..., Any]) -> AssetsDefinition:
+        return _GraphBackedAsset(
+            name=cast(Optional[str], name),  # (mypy bug that it can't infer name is Optional[str])
+            key_prefix=key_prefix,
+            ins=ins,
+            description=description,
+            partitions_def=partitions_def,
+            group_name=group_name,
+        )(fn)
+
+    return inner
+
+
+class _GraphBackedAsset:
+    def __init__(
+        self,
+        name: Optional[str] = None,
+        key_prefix: Optional[CoercibleToAssetKeyPrefix] = None,
+        ins: Optional[Mapping[str, AssetIn]] = None,
+        description: Optional[str] = None,
+        partitions_def: Optional[PartitionsDefinition] = None,
+        group_name: Optional[str] = None,
+    ):
+        self.name = name
+
+        if isinstance(key_prefix, str):
+            key_prefix = [key_prefix]
+        self.key_prefix = key_prefix
+        self.ins = ins or {}
+        self.description = description
+        self.partitions_def = partitions_def
+        self.group_name = group_name
+
+    def __call__(self, fn: Callable) -> AssetsDefinition:
+        asset_name = self.name or fn.__name__
+        asset_ins = build_asset_ins(fn, self.ins or {}, set())
+        out_asset_key = AssetKey(list(filter(None, [*(self.key_prefix or []), asset_name])))
+
+        keys_by_input_name = {
+            input_name: asset_key for asset_key, (input_name, _) in asset_ins.items()
+        }
+        partition_mappings = {
+            input_name: asset_in.partition_mapping
+            for input_name, asset_in in self.ins.items()
+            if asset_in.partition_mapping
+        }
+
+        op_graph = graph(
+            name="__".join(out_asset_key.path).replace("-", "_"), description=self.description
+        )(fn)
+        return AssetsDefinition.from_graph(
+            op_graph,
+            keys_by_input_name=keys_by_input_name,
+            keys_by_output_name={"result": out_asset_key},
+            partitions_def=self.partitions_def,
+            partition_mappings=partition_mappings if partition_mappings else None,
+            group_name=self.group_name,
+        )
+
+
+def graph_multi_asset(
+    *,
+    outs: Mapping[str, AssetOut],
+    name: Optional[str] = None,
+    ins: Optional[Mapping[str, AssetIn]] = None,
```

Comment:
Can happen in a follow up, but we likely want to support `internal_asset_deps` here

Reply:
Very true